kernel/kernel-defaults/src/test/scala/io/delta/kernel/defaults/CheckpointV2ReadSuite.scala

```diff
@@ -0,0 +1,213 @@
+            spark.sql(s"INSERT INTO $tbl VALUES (1, 'a'), (2, 'b')")
+            spark.sql(s"INSERT INTO $tbl VALUES (3, 'c'), (4, 'd')")
+            spark.sql(s"INSERT INTO $tbl VALUES (5, 'e'), (6, 'f')")
+          }
+
+          // Validate snapshot and data.
+          validateSnapshot(path.toString, DeltaLog.forTable(spark, path.toString).update())
+          checkTable(
+            path = path.toString,
+            expectedAnswer = (1 to 6).map(i => TestRow(i, (i - 1 + 'a').toChar.toString))
+          )
+
+          // Remove some files from the table, then add a new one.
+          spark.sql(s"DELETE FROM $tbl WHERE a=1 OR a=2")
+          spark.sql(s"INSERT INTO $tbl VALUES (7, 'g'), (8, 'h')")
+
+          // Validate snapshot and data.
+          validateSnapshot(path.toString, DeltaLog.forTable(spark, path.toString).update())
+          checkTable(
+            path = path.toString,
+            expectedAnswer = (3 to 8).map(i => TestRow(i, (i - 1 + 'a').toChar.toString))
+          )
+        }
+      }
+    }
+  }
+
+  test("UUID named checkpoint with actions") {
+    withTempDir { path =>
+      // Create Delta log and a checkpoint file with actions in it.
+      val log = DeltaLog.forTable(spark, new Path(path.toString))
+      new File(log.logPath.toUri).mkdirs()
+
+      val metadata = Metadata("testId", schemaString = "{\"type\":\"struct\",\"fields\":[" +
+        "{\"name\":\"a\",\"type\":\"integer\",\"nullable\":true,\"metadata\":{}}]}")
+      val supportedFeatures = Set("v2Checkpoint", "appendOnly", "invariants")
+      val protocol = Protocol(3, 7, Some(Set("v2Checkpoint")), Some(supportedFeatures))
+      val add = AddFile(new Path("addfile").toUri.toString, Map.empty, 100L,
+        10L, dataChange = true)
+
+      log.startTransaction().commitManually(Seq(metadata, add): _*)
+      log.upgradeProtocol(None, log.update(), protocol)
+      log.checkpoint(log.update())
+
+      // Spark snapshot and files must be evaluated before renaming the checkpoint file.
+      // This is because this checkpoint file (technically) becomes invalid, as there is no
+      // CheckpointManifest action in it. However, because the Spark connector will place all
+      // Add and Remove actions in the sidecar files, we must use this hack to test this
+      // scenario.
+      val snapshotFromSpark = DeltaLog.forTable(spark, path.toString).update()
+      snapshotFromSpark.allFiles.collect()
+
+      // Rename to UUID.
+      val ckptPath = new Path(new File(log.logPath.toUri).listFiles().filter(f =>
+        FileNames.isCheckpointFile(new Path(f.getPath))).head.toURI)
+      new File(ckptPath.toUri).renameTo(new File(new Path(ckptPath.getParent, ckptPath.getName
+        .replace("checkpoint.parquet", "checkpoint.abc-def.parquet")).toUri))
+
+      // Validate snapshot.
+      validateSnapshot(path.toString, snapshotFromSpark)
+    }
+  }
+
+  test("compatibility checkpoint with sidecar files") {
```

Comment:
Can we also have a test case where both checkpoints - v2 and compat are present and assert that v2 is picked.

Reply:
Lets add here also. From the `SnapshotImpl` you can get the LogSegment and verify it.

Reply:
@chirag-s-db Is this handled?